core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1327,69 +1328,337 @@ class ReassignedPartitionsIsrChangeListener(protected val controller: KafkaContr
+        !controllerContext.liveOrShuttingDownBrokerIds.contains(p._2.leaderAndIsr.leader) &&
+          !topicDeletionManager.isTopicQueuedUpForDeletion(p._1.topic)
+      )
+    offlinePartitionCount.set(opc)
+
+    val pric = if (!isActive)
+      0
+    else
+      controllerContext.partitionReplicaAssignment.count { case (topicPartition, replicas) =>
+        controllerContext.partitionLeadershipInfo.contains(topicPartition) &&
+        controllerContext.partitionLeadershipInfo(topicPartition).leaderAndIsr.leader != replicas.head &&
+          !topicDeletionManager.isTopicQueuedUpForDeletion(topicPartition.topic)
+      }
+    preferredReplicaImbalanceCount.set(pric)
+  }
+
+  private def triggerControllerMove(): Unit = {
+    activeControllerId.set(-1)
+    controllerContext.zkUtils.deletePath(ZkUtils.ControllerPath)
+  }
+
+  def elect(): Unit = {
+    val timestamp = time.milliseconds
+    val electString = ZkUtils.controllerZkData(config.brokerId, timestamp)
+
+    activeControllerId.set(getControllerID())
+    /*
+     * We can get here during the initial startup and the handleDeleted ZK callback. Because of the potential race condition,
+     * it's possible that the controller has already been elected when we get here. This check will prevent the following
+     * createEphemeralPath method from getting into an infinite loop if this broker is already the controller.
+     */
+    if(activeControllerId.get() != -1) {
+      debug("Broker %d has been elected as leader, so stopping the election process.".format(activeControllerId.get()))
+      return
+    }
+
+    try {
+      val zkCheckedEphemeral = new ZKCheckedEphemeral(ZkUtils.ControllerPath,
+                                                      electString,
+                                                      controllerContext.zkUtils.zkConnection.getZookeeper,
+                                                      controllerContext.zkUtils.isSecure)
+      zkCheckedEphemeral.create()
+      info(config.brokerId + " successfully elected as leader")
+      activeControllerId.set(config.brokerId)
+      onControllerFailover()
+    } catch {
+      case _: ZkNodeExistsException =>
+        // If someone else has written the path, then
+        activeControllerId.set(getControllerID)
+
+        if (activeControllerId.get() != -1)
+          debug("Broker %d was elected as leader instead of broker %d".format(activeControllerId.get(), config.brokerId))
+        else
+          warn("A leader has been elected but just resigned, this will result in another round of election")
+
+      case e2: Throwable =>
+        error("Error while electing or becoming leader on broker %d".format(config.brokerId), e2)
+        triggerControllerMove()
+    }
+  }
+}
+
+/**
+  * This is the zookeeper listener that triggers all the state transitions for a replica
+  */
+class BrokerChangeListener(protected val controller: KafkaController) extends IZkChildListener with Logging {
```

Comment:
The `protected val` was an artifact of the existing code which I forgot to remove. Thanks for pointing it out!